base/src/H264DecoderV4L2Helper.cpp

```diff
@@ -0,0 +1,1478 @@
+        if (ctx->cp_mem_type == V4L2_MEMORY_DMABUF)
+        {
+            // For NV12 format.
+            v4l2_buf.m.planes[0].m.fd = ctx->dmabuff_fd[i];
+            v4l2_buf.m.planes[1].m.fd = ctx->dmabuff_fd[i];
+        }
+ 
+        ret_val = q_buffer(ctx, v4l2_buf, NULL, ctx->cp_buf_type, ctx->cp_mem_type,
+            ctx->cp_num_planes);
+ 
+        if (ret_val)
+        {
+            LOG_ERROR << "Qing failed on capture plane" << endl;
+            ctx->in_error = 1;
+            return ;
+        }
+    }
+ 
+    // Set streaming status ON on capture plane.
+ 
+    ret_val = v4l2_ioctl(ctx->fd,VIDIOC_STREAMON, &ctx->cp_buf_type);
+    if (ret_val != 0)
+    {
+        LOG_ERROR << "Streaming error on capture plane" << endl;
+        ctx->in_error = 1;
+    }
+    ctx->cp_streamon = 1;
+ 
+    LOG_INFO << "Query and set capture successful" << endl;
+ 
+    return;
+}
+ 
+void * h264DecoderV4L2Helper::capture_thread(void *arg) 
+{
+    h264DecoderV4L2Helper *m_nThread = (h264DecoderV4L2Helper*)arg;
+    context_t* ctx = &m_nThread->ctx;
+    struct v4l2_event event;
+    int ret_val;
+ 
+    /* Need to wait for the first Resolution change event, so that
+    ** the decoder knows the stream resolution and can allocate
+    ** appropriate buffers when REQBUFS is called.
+    *///rvw
+ 
+    do
+    {
+        // Dequeue the subscribed event.
+        ret_val = m_nThread->dq_event(ctx, event, 5000);
+        if (ret_val)
+        {
+            if (errno == EAGAIN)
+            {
+                LOG_ERROR << "Timeout waiting for first V4L2_EVENT_RESOLUTION_CHANGE" << endl;
+            }
+            else
+            {
+                LOG_ERROR << "Error in dequeueing decoder event" << endl;
+            }
+            ctx->in_error = 1;
+            break;
+        }
+    }
+    while ((event.type != V4L2_EVENT_RESOLUTION_CHANGE) && !ctx->in_error);
+ 
+    /* Recieved first resolution change event
+    ** Format and buffers are now set on capture.
+    */
+
+    auto outputFrame = m_nThread->makeFrame();
+    auto dmaOutFrame = static_cast<DMAFDWrapper *>(outputFrame->data());
+    int f_d = dmaOutFrame->getFd();
+
+    if (!ctx->in_error)
+    {
+       m_nThread->query_set_capture(ctx, f_d);
+    }
+ 
+    /* Check for resolution event to again
+    ** set format and buffers on capture plane.
+    */
+    while (!(ctx->in_error || ctx->got_eos))
+    {
+        Buffer *decoded_buffer = new Buffer(ctx->cp_buf_type, ctx->cp_mem_type, 0);
+         ret_val = m_nThread->dq_event(ctx, event, 0);
+        if (ret_val == 0)
+        {
+            switch (event.type)
+            {
+                case V4L2_EVENT_RESOLUTION_CHANGE:
+                    m_nThread->query_set_capture(ctx, f_d);
+                    continue; 
+            }
+        }
+        // Main Capture loop for DQ and Q.
+ 
+        while (1)
+        {
+            Buffer *decoded_buffer = new Buffer(ctx->cp_buf_type, ctx->cp_mem_type, 0);
```

Comment:
Who is deleting this Buffer ? it is being allocated in a loop ? is ti a memory leak ?